tests/cover/test_complex_numbers.py

```diff
@@ -0,0 +1,126 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2018 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import sys
+import math
+
+import hypothesis.strategies as st
+from hypothesis import given, reject
+from tests.common.debug import minimal
+from hypothesis.strategies import complex_numbers
+
+
+def test_minimal():
+    assert minimal(complex_numbers(), lambda x: True) == 0
+
+
+def test_minimal_nonzero_real():
+    assert minimal(complex_numbers(), lambda x: x.real != 0) == 1
+
+
+def test_minimal_nonzero_imaginary():
+    assert minimal(complex_numbers(), lambda x: x.imag != 0) == 1j
+
+
+def test_minimal_quadrant1():
+    assert minimal(
+        complex_numbers(),
+        lambda x: x.imag > 0 and x.real > 0
+    ) == 1 + 1j
+
+
+def test_minimal_quadrant2():
+    assert minimal(
+        complex_numbers(),
+        lambda x: x.imag > 0 and x.real < 0
+    ) == -1 + 1j
+
+
+def test_minimal_quadrant3():
+    assert minimal(
+        complex_numbers(),
+        lambda x: x.imag < 0 and x.real < 0
+    ) == -1 - 1j
+
+
+def test_minimal_quadrant4():
+    assert minimal(
+        complex_numbers(),
+        lambda x: x.imag < 0 and x.real > 0
+    ) == 1 - 1j
+
+
+@given(st.data(), st.integers(-5, 5).map(lambda x: 10 ** x))
+def test_max_magnitude_respected(data, mag):
+    c = data.draw(complex_numbers(max_magnitude=mag))
+    assert abs(c) <= mag * (1 + sys.float_info.epsilon)
+
+
+@given(complex_numbers(max_magnitude=0))
+def test_max_magnitude_zero(val):
+    assert val == 0
+
+
+@given(st.data(), st.integers(-5, 5).map(lambda x: 10 ** x))
+def test_min_magnitude_respected(data, mag):
+    c = data.draw(complex_numbers(min_magnitude=mag))
+    assert (
+        abs(c.real) >= mag or
+        abs(c.imag) >= mag or
+        abs(c) >= mag * (1 - sys.float_info.epsilon)
+    )
+
+
+def test_minimal_min_magnitude_zero():
+    assert minimal(
+        complex_numbers(min_magnitude=0),
+        lambda x: True
+    ) == 0
+
+
+def test_minimal_min_magnitude_none():
+    assert minimal(
+        complex_numbers(min_magnitude=None),
+        lambda x: True
+    ) == 0
+
+
+def test_minimal_min_magnitude_positive():
+    assert minimal(
+        complex_numbers(min_magnitude=0.5),
+        lambda x: True
+    ) in (0.5, 1)
```

Comment:
I think that in this case, the "minimal" results should be is 1+0j (I found that replacing 0.5 by any other number strictly between 0 and 1, that is the "minimal" value found).  Does this point to an issue with the `@composite` approach? i.e., once the first component is minimised, the subsequent components are not?

Reply:
Subsequent components definitely are minimized (if they weren't you'd see a lot more variation than this!), but it may be that something weird is happening around the 0.5 boundary - powers of two are a special case in floating point minimization because they're a special case in floating point representation.

In general a thing that can go wrong in shrinking with composite is that when there are dependencies between generated values it may be the case that the shrink that would be required is not one that the underlying shrinker can find.

My personal recommendation is don't sweat it for this release (we've shipped *much* worse floating point shrinking than this!) and that if you want to dig into this I'd recommend a follow-up pull request digging into shrinker internals, which I'd be very happy to help you with!

Reply:
I did try a case with `min_magnitude` something which is "as far as minimal as I could imagine", 0.4374632746 or such-like, and that was the result; IIRC I switched that to 0.5 for the commit (I'm not on my dev machine now so can't reproduce).

But agreed, better bumped to an issue rather than holding up the PR.

Reply:
> My personal recommendation is don't sweat it for this release (we've shipped much worse floating point shrinking than this!)

That is exactly my plan - hence writing the tests to pass with 0.5+0j or 1+0j.  In fact the `in (0.5, 1)` assertion was copied of an old floating-point shrinking test which we seem to have since removed :smile: